core/services/head_tracker.go

```diff
@@ -277,32 +282,119 @@ func (ht *HeadTracker) handleNewHead(bh gethTypes.Header) error {
 	}
 	if head.Number == prevHead.Number {
 		if head.Hash != prevHead.Hash {
-			logger.Debugf("duplicate blocks at height %v. Got block hash %s but already saw block hash %s", head.Number, head.Hash.Hex(), ht.highestSeenHead.Hash.Hex())
+			logger.Debugw("HeadTracker: got duplicate head", "blockNum", head.Number, "gotHead", head.Hash.Hex(), "highestSeenHead", ht.highestSeenHead.Hash.Hex())
 		} else {
-			logger.Debugf("head with hash %s was already in the database", head.Hash.Hex())
+			logger.Debugw("HeadTracker: head already in the database", "gotHead", head.Hash.Hex())
 		}
 	} else {
-		logger.Debugf("received out of order head %s with number %v. Latest head is at %v", head.Hash.Hex(), head.Number, ht.highestSeenHead.Number)
+		logger.Debugw("HeadTracker: got out of order head", "blockNum", head.Number, "gotHead", head.Hash.Hex(), "highestSeenHead", ht.highestSeenHead.Number)
 	}
 	return nil
 }
 
 func (ht *HeadTracker) handleNewHighestHead(head models.Head) error {
-	headWithChain, err := ht.store.Chain(head.Hash, ht.store.Config.EthFinalityDepth())
+	// NOTE: We must set a hard time limit on this, backfilling heads should
+	// not block the head tracker
+	ctx, cancel := context.WithTimeout(context.Background(), backfillTimeBudget)
```

Comment:
There should really be a single background context for the entire process. That should get propagated and specialized throughout the components of the node.

Eg. this process queries a Geth node for blocks and stores metadata in the DB for the missing ones. The DB and HTTP requests should have this context attached to properly handle the cancellation. 

Afaict, it's only the Ethereum client that supports contexts.